kratos/utilities/entities_utilities.cpp

```diff
@@ -12,16 +12,199 @@
 
 // Project includes
 #include "utilities/entities_utilities.h"
+#include "utilities/string_utilities.h"
+#include "utilities/geometry_utilities.h"
 
-namespace Kratos
+namespace Kratos::EntitiesUtilities
 {
-namespace EntitiesUtilities
+
+template<class TEntity>
+EntitityIdentifier<TEntity>::EntitityIdentifier(const std::string& rName)
+{
+    // Check if the name is empty (could be empty due to default values in settings)
+    if (rName != "") {
+        // Identify type of definition
+        if (rName.find("#") != std::string::npos) { // Templated definition
+            if (rName.find(";") == std::string::npos) { // Must be single definition
+                GenerateTemplatedTypes(rName);
+            } else {
+                KRATOS_ERROR << "Unsupported definition type. Cannot use # and ; together" << std::endl;
+            }
+        } else { // Multiple definition or single definition
+            if (rName.find(";") == std::string::npos) { // Must be single definition
+                GenerateSingleType(rName);
+            } else {
+                GenerateMultipleTypes(rName);
+            }
+        }
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<class TEntity>
+bool EntitityIdentifier<TEntity>::IsInitialized()
+{
+    if (mpPrototypeEntity == nullptr) {
+        return false;
+    } else {
+        // Check if the class is initialized
+        if (mDefinitionType == DefinitionType::Multiple) {
+            // Check size 
+            if (mTypes.size() > 0) {
+                return true;
+            } else {
+                return false;
+            }
+        } else { // Directly check the name
+            return true;
+        }
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<class TEntity>
+const TEntity& EntitityIdentifier<TEntity>::GetPrototypeEntity(typename GeometryType::Pointer pGeometry)
+{
+    switch (mDefinitionType) {
+        case DefinitionType::Single:
+            KRATOS_DEBUG_ERROR_IF_NOT(pGeometry->GetGeometryType() == mpPrototypeEntity->GetGeometry().GetGeometryType()) << "Trying to replace an entity with a different geometry type. Reference entity " << mpPrototypeEntity->GetGeometry().Info() << " vs  " << pGeometry->Info() << "\n Entity info: " << mpPrototypeEntity->Info() << std::endl;
+            return *mpPrototypeEntity;
+        case DefinitionType::Multiple:
+            // We check if the current type is correct or retrieving is required
+            if (pGeometry->GetGeometryType() != mpPrototypeEntity->GetGeometry().GetGeometryType()) {
+                mpPrototypeEntity = (mTypes[pGeometry->GetGeometryType()])->Create(0, pGeometry, nullptr);
+            }
+            return *mpPrototypeEntity;
```

Comment:
here you should return 

mTypes[pGeometry->GetGeometryType()])->second 

mpPrototypeEntity should never be used in DefinitionType::Multiple

Reply:
Read my previous message, please, this is needed to minimize costs